torch/_mock_dispatcher/dispatch_key.py

```diff
@@ -1,47 +1,96 @@
 from enum import Enum, auto
-from typing import List
+from typing import List, Iterator
 
 class DispatchKey(Enum):
-    Undefined = 0
+    # "Dense" backends
+    # These are "real" backends that can also have different sparse/autograd/quantized kernels
+    CPUBit = 0
+    CUDABit = auto()
+    HIPBit = auto()
+    XLABit = auto()
+    LazyBit = auto()
+    XPUBit = auto()
+    NestedTensorBit = auto()
+    MLCBit = auto()
+    HPUBit = auto()
+    PrivateUse1Bit = auto()
+    PrivateUse2Bit = auto()
+    PrivateUse3Bit = auto()
+
+    # EndOfExtensibleBackendKeys tells us which keys should get "backend bits" in the runtime bitset.
+    EndOfBackendKeys = PrivateUse3Bit
+
+    # Undefined technically counts as a "functionality" bit - it's the lowest priority functionality.
+    Undefined = auto()
     CatchAll = Undefined
 
-    CPU = auto()
-    CUDA = auto()
-    HIP = auto()
+
+    # Dense lives above the "Other" backends, mostly because the other backends should have higher priority than dense CPU.
+    # e.g. torch.add(vulkan_tensor, cpu_tensor) should dispatch to the vulkan kernel
+    Dense = auto()
+
```

Comment:
> This is the logic that needs to go from "bitset representation" -> "index into the operator table".
You can see the details in the PR (at the bottom of dispatch_key_set.py), but the new logic will definitely be slower. It currently consists of:

Is there an explanation of what happens today (how one goes from a DispatchKeySet(is this the right analogy?) to an index into the operator table)? Do we have some idea of how much additional performance overhead that will be right now?

Reply:
> Is there an explanation of what happens today for comparison

Yep- for comparison, today is just a single `highestPriorityTypeId()` call. This change will add in a handful of bitshifting operations, which will definitely be slower. The hope is that bitshifting operations are generally very fast - probably faster than, say, an extra redispatch (which was one of the proposed alternatives originally).